### DISCUSSION TITLE
Are you using Hera?

### DISCUSSION THREAD

**Discussion by flaviuvadan** [argoproj-labs/hera#625] [General]
We always love hearing from the community! If you are using Hera we'd love to hear:
- what problems you use Hera and Argo Workflows for in your organization
- whether you or your users have any feedback for Hera

If you are interested in collaborating on a blog post to promote your organization and Hera, we'd love to hear from you! 

#### Lastly, if you are open to it, please submit a PR to add your organization to [Hera's users list](https://github.com/argoproj-labs/hera/blob/main/USERS.md)!

**Comment by Axolotling** [comment]
Hi! I'm new to hera. Recently I'm trying to apply it to orchestrate statistical model training. 

The idea to write python code instead of yamls seems great, although I have some constructive feedback on the execution. 

The main challenge I'm facing is that I can't seem to find any complex examples online. It's mostly simple examples like the 2 ones in Readme. 
This makes it very hard to actually start developing something serious in hera, since I have to guess which of the api elements interact with which other ones in what way (and each element has many optional parameters, which makes guessing even harder). 

Hera could gain more traction if there were various complex examples for new users to learn from.

**Comment by flaviuvadan** [comment]
Thanks so much for the feedback @Axolotling! I agree we do need more complex examples. Any chance you have some suggestions on what would help you most for onboarding purposes? I can prioritize writing something up and pushing it as an example. Perhaps something with MNIST, some VAEs hyperparameter grid search  + choosing a model based on loss might be useful? Happy to provide the examples

**Reply by Axolotling** [reply]
To be honest the subject of tasks solved in the examples does not seem important. Maybe even easy toy-tasks would be better. The complexity that would be helpful from my perspective would be using as many features of argo workflows (through Hera) as possible. 

Let me present some specific examples that would be useful to me:
A) WorkflowTemplate that is invoked as one of steps of other WorkflowTemplate
B) Same as A, but with result passed as parameter from one step to the next step and used there for some calculation
C) Same as B, but with a parallel execution per each parameter from list passed from previous step
D) Same as B, but with more complex sets of params (basically more than one variable)
E) Same as D, but using `with_items` or `with_param` to showcase what is the difference between them

For the context I'd like to work with WorkflowTemplates because they can be both: 
- referenced from other Workflows as a part of logic
- invoked manually with a set of params as needed

This was an example of things useful specifically to me, but I imagine people would like to implement basically any behavior of argo-workflows in Hera, so analogous examples but eg. with Workflows instead of WorkflowTemplates would also be useful for somebody. 
The key idea here is to show how the library can be used, not necessarily for what can it be used.

**Reply by flaviuvadan** [reply]
Thanks for outlining all those examples! That's very helpful! 

I must ask - any chance you saw the [Hera examples page](https://hera.readthedocs.io/en/stable/examples/workflows/artifact_with_fanout/)? There's a section of Hera examples that primarily illustrates features of Hera proper, and then a section that reproduces a lot of the [upstream examples from Argo Workflows but in Hera](https://hera.readthedocs.io/en/stable/examples/workflows/upstream/cluster_workflow_template__cluster_wftmpl_dag/)? The one I linked, for instance, illustrates a template reference in step D, which is also parameterized, which can be combined with an [outputs example](https://hera.readthedocs.io/en/stable/examples/workflows/dag_with_script_output_param_passing/). I wonder whether you have any feedback on those ones?

**Reply by Axolotling** [reply]
Oh, that seems very helpful. A little embarassing that I was looking so hard for examples and did not notice that page 😅. 
Thank you, all I can say now is good job on having these examples!

**Reply by Axolotling** [reply]
I guess there is one additional piece of feedback I can give. Consider improving the visibility of the examples. I have seen that page before, but assumed there is only the list of missing examples and nothing more there. 

The side menu only said "Hera" (it was not expanded by default), so it did not make it seem that there would be examples when I click that.
<img width="325" alt="image" src="https://github.com/argoproj-labs/hera/assets/22130690/80faa469-9631-4cbd-bcad-1bc418ba94b3">

Idea 1. Rename the side menu entry to "Hera Examples" or something similar, and have it expanded by default.
Idea 2. Make the list of examples be the content of the main page of "Examples" in place of the list of missing examples, missing examples can be less visible than the actual examples.

**Reply by flaviuvadan** [reply]
I am incredibly grateful for all the feedback and thought you've put into writing this all out @Axolotling! I opened a documentation issue [here](https://github.com/argoproj-labs/hera/issues/733) and will leave this unresolved until we incorporate some of your thoughts into the docs. I'd love to hear your other thoughts on Hera as well :)
